File: packages/jumpboot-js/msgpack.js
"use strict";
// Minimal, dependency-free MessagePack codec for the jumpboot queue protocol.
//
// It implements the subset of the MessagePack spec the protocol uses — nil,
// bool, int, float64, str, bin, array, map — and interoperates with the Go
// host's github.com/vmihailenco/msgpack/v5. It is bundled (not npm-installed)
// so a Node.js QueueProcess works the moment the environment is created, the
// same way the Python package bundles a pure-Python msgpack fallback.
//
// See docs/PROTOCOL.md for the framing that wraps these payloads.

// ---- encode ---------------------------------------------------------------

function encode(value) {
	const chunks = [];
	encodeValue(value, chunks);
	return Buffer.concat(chunks);
}

function encodeValue(v, out) {
	if (v === null || v === undefined) {
		out.push(Buffer.from([0xc0]));
		return;
	}
	switch (typeof v) {
		case "boolean":
			out.push(Buffer.from([v ? 0xc3 : 0xc2]));
			return;
		case "number":
			encodeNumber(v, out);
			return;
		case "string":
			encodeString(v, out);
			return;
		case "object":
			if (Buffer.isBuffer(v) || v instanceof Uint8Array) {
				encodeBinary(Buffer.from(v), out);
			} else if (Array.isArray(v)) {
				encodeArray(v, out);
			} else {
				encodeMap(v, out);
			}
			return;
		default:
			throw new TypeError("msgpack: cannot encode value of type " + typeof v);
	}
}

function encodeNumber(n, out) {
	if (Number.isInteger(n)) {
		if (n >= 0) {
			if (n < 0x80) {
				out.push(Buffer.from([n]));
			} else if (n < 0x100) {
				out.push(Buffer.from([0xcc, n]));
			} else if (n < 0x10000) {
				const b = Buffer.alloc(3);
				b[0] = 0xcd;
				b.writeUInt16BE(n, 1);
				out.push(b);
			} else if (n < 0x100000000) {
				const b = Buffer.alloc(5);
				b[0] = 0xce;
				b.writeUInt32BE(n, 1);
				out.push(b);
			} else {
				const b = Buffer.alloc(9);
				b[0] = 0xcf;
				b.writeBigUInt64BE(BigInt(n), 1);
				out.push(b);
			}
		} else {
			if (n >= -0x20) {
				out.push(Buffer.from([n & 0xff]));
			} else if (n >= -0x80) {
				out.push(Buffer.from([0xd0, n & 0xff]));
			} else if (n >= -0x8000) {
				const b = Buffer.alloc(3);
				b[0] = 0xd1;
				b.writeInt16BE(n, 1);
				out.push(b);
			} else if (n >= -0x80000000) {
				const b = Buffer.alloc(5);
				b[0] = 0xd2;
				b.writeInt32BE(n, 1);
				out.push(b);
			} else {
				const b = Buffer.alloc(9);
				b[0] = 0xd3;
				b.writeBigInt64BE(BigInt(n), 1);
				out.push(b);
			}
		}
		return;
	}
	// Non-integer: encode as float64.
	const b = Buffer.alloc(9);
	b[0] = 0xcb;
	b.writeDoubleBE(n, 1);
	out.push(b);
}

function encodeString(s, out) {
	const data = Buffer.from(s, "utf8");
	const len = data.length;
	if (len < 0x20) {
		out.push(Buffer.from([0xa0 | len]));
	} else if (len < 0x100) {
		out.push(Buffer.from([0xd9, len]));
	} else if (len < 0x10000) {
		const h = Buffer.alloc(3);
		h[0] = 0xda;
		h.writeUInt16BE(len, 1);
		out.push(h);
	} else {
		const h = Buffer.alloc(5);
		h[0] = 0xdb;
		h.writeUInt32BE(len, 1);
		out.push(h);
	}
	out.push(data);
}

function encodeBinary(data, out) {
	const len = data.length;
	if (len < 0x100) {
		out.push(Buffer.from([0xc4, len]));
	} else if (len < 0x10000) {
		const h = Buffer.alloc(3);
		h[0] = 0xc5;
		h.writeUInt16BE(len, 1);
		out.push(h);
	} else {
		const h = Buffer.alloc(5);
		h[0] = 0xc6;
		h.writeUInt32BE(len, 1);
		out.push(h);
	}
	out.push(data);
}

function encodeArray(arr, out) {
	const len = arr.length;
	if (len < 0x10) {
		out.push(Buffer.from([0x90 | len]));
	} else if (len < 0x10000) {
		const h = Buffer.alloc(3);
		h[0] = 0xdc;
		h.writeUInt16BE(len, 1);
		out.push(h);
	} else {
		const h = Buffer.alloc(5);
		h[0] = 0xdd;
		h.writeUInt32BE(len, 1);
		out.push(h);
	}
	for (const item of arr) encodeValue(item, out);
}

function encodeMap(obj, out) {
	const keys = Object.keys(obj);
	const len = keys.length;
	if (len < 0x10) {
		out.push(Buffer.from([0x80 | len]));
	} else if (len < 0x10000) {
		const h = Buffer.alloc(3);
		h[0] = 0xde;
		h.writeUInt16BE(len, 1);
		out.push(h);
	} else {
		const h = Buffer.alloc(5);
		h[0] = 0xdf;
		h.writeUInt32BE(len, 1);
		out.push(h);
	}
	for (const key of keys) {
		encodeString(key, out);
		encodeValue(obj[key], out);
	}
}

// ---- decode ---------------------------------------------------------------

// decode reads exactly one MessagePack value from buf and returns it. The
// caller frames messages by length (see transport.js), so trailing bytes are
// not expected.
function decode(buf) {
	const reader = { buf, pos: 0 };
	return decodeValue(reader);
}

function decodeValue(r) {
	const b = r.buf[r.pos++];

	if (b <= 0x7f) return b; // positive fixint
	if (b >= 0xe0) return b - 0x100; // negative fixint
	if (b >= 0x80 && b <= 0x8f) return decodeMap(r, b & 0x0f); // fixmap
	if (b >= 0x90 && b <= 0x9f) return decodeArray(r, b & 0x0f); // fixarray
	if (b >= 0xa0 && b <= 0xbf) return decodeStr(r, b & 0x1f); // fixstr

	switch (b) {
		case 0xc0:
			return null;
		case 0xc2:
			return false;
		case 0xc3:
			return true;
		case 0xc4:
			return decodeBin(r, readUint(r, 1));
		case 0xc5:
			return decodeBin(r, readUint(r, 2));
		case 0xc6:
			return decodeBin(r, readUint(r, 4));
		case 0xca: {
			const v = r.buf.readFloatBE(r.pos);
			r.pos += 4;
			return v;
		}
		case 0xcb: {
			const v = r.buf.readDoubleBE(r.pos);
			r.pos += 8;
			return v;
		}
		case 0xcc:
			return readUint(r, 1);
		case 0xcd:
			return readUint(r, 2);
		case 0xce:
			return readUint(r, 4);
		case 0xcf:
			return readUint(r, 8);
		case 0xd0:
			return readInt(r, 1);
		case 0xd1:
			return readInt(r, 2);
		case 0xd2:
			return readInt(r, 4);
		case 0xd3:
			return readInt(r, 8);
		case 0xd9:
			return decodeStr(r, readUint(r, 1));
		case 0xda:
			return decodeStr(r, readUint(r, 2));
		case 0xdb:
			return decodeStr(r, readUint(r, 4));
		case 0xdc:
			return decodeArray(r, readUint(r, 2));
		case 0xdd:
			return decodeArray(r, readUint(r, 4));
		case 0xde:
			return decodeMap(r, readUint(r, 2));
		case 0xdf:
			return decodeMap(r, readUint(r, 4));
		default:
			throw new Error("msgpack: unsupported type byte 0x" + b.toString(16));
	}
}

function readUint(r, n) {
	let v;
	if (n === 1) {
		v = r.buf[r.pos];
	} else if (n === 2) {
		v = r.buf.readUInt16BE(r.pos);
	} else if (n === 4) {
		v = r.buf.readUInt32BE(r.pos);
	} else {
		// 8 bytes — collapse to a JS number (safe for our protocol's values).
		v = Number(r.buf.readBigUInt64BE(r.pos));
	}
	r.pos += n;
	return v;
}

function readInt(r, n) {
	let v;
	if (n === 1) {
		v = r.buf.readInt8(r.pos);
	} else if (n === 2) {
		v = r.buf.readInt16BE(r.pos);
	} else if (n === 4) {
		v = r.buf.readInt32BE(r.pos);
	} else {
		v = Number(r.buf.readBigInt64BE(r.pos));
	}
	r.pos += n;
	return v;
}

function decodeStr(r, len) {
	const s = r.buf.toString("utf8", r.pos, r.pos + len);
	r.pos += len;
	return s;
}

function decodeBin(r, len) {
	const b = r.buf.subarray(r.pos, r.pos + len);
	r.pos += len;
	return Buffer.from(b);
}

function decodeArray(r, len) {
	const arr = new Array(len);
	for (let i = 0; i < len; i++) arr[i] = decodeValue(r);
	return arr;
}

function decodeMap(r, len) {
	const obj = {};
	for (let i = 0; i < len; i++) {
		const key = decodeValue(r);
		obj[key] = decodeValue(r);
	}
	return obj;
}

module.exports = { encode, decode };

File: wasmguest/msgpack.go
package wasmguest

import (
	"encoding/binary"
	"fmt"
	"math"
)

// Minimal, stdlib-only MessagePack codec for the jumpboot queue protocol.
// It implements the subset the protocol uses — nil, bool, int, float, str,
// bin, array, map — and interoperates with the Go host's
// github.com/vmihailenco/msgpack/v5. Decoded integers are returned as int64,
// floats as float64, strings as string, maps as map[string]any.

// ---- encode ---------------------------------------------------------------

func mpEncode(v any) []byte {
	var out []byte
	return mpEncodeValue(out, v)
}

func mpEncodeValue(out []byte, v any) []byte {
	switch x := v.(type) {
	case nil:
		return append(out, 0xc0)
	case bool:
		if x {
			return append(out, 0xc3)
		}
		return append(out, 0xc2)
	case int:
		return mpEncodeInt(out, int64(x))
	case int8:
		return mpEncodeInt(out, int64(x))
	case int16:
		return mpEncodeInt(out, int64(x))
	case int32:
		return mpEncodeInt(out, int64(x))
	case int64:
		return mpEncodeInt(out, x)
	case uint:
		return mpEncodeUint(out, uint64(x))
	case uint8:
		return mpEncodeUint(out, uint64(x))
	case uint16:
		return mpEncodeUint(out, uint64(x))
	case uint32:
		return mpEncodeUint(out, uint64(x))
	case uint64:
		return mpEncodeUint(out, x)
	case float32:
		return mpEncodeFloat(out, float64(x))
	case float64:
		return mpEncodeFloat(out, x)
	case string:
		return mpEncodeString(out, x)
	case []byte:
		return mpEncodeBinary(out, x)
	case []any:
		return mpEncodeArray(out, x)
	case map[string]any:
		return mpEncodeMap(out, x)
	default:
		// Unsupported types are encoded as their fmt.Sprint form so a stray
		// value can never wedge the protocol loop.
		return mpEncodeString(out, fmt.Sprint(x))
	}
}

func mpEncodeInt(out []byte, n int64) []byte {
	if n >= 0 {
		return mpEncodeUint(out, uint64(n))
	}
	switch {
	case n >= -0x20:
		return append(out, byte(n))
	case n >= -0x80:
		return append(out, 0xd0, byte(n))
	case n >= -0x8000:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		return append(out, 0xd1, b[0], b[1])
	case n >= -0x80000000:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		return append(append(out, 0xd2), b[:]...)
	default:
		var b [8]byte
		binary.BigEndian.PutUint64(b[:], uint64(n))
		return append(append(out, 0xd3), b[:]...)
	}
}

func mpEncodeUint(out []byte, n uint64) []byte {
	switch {
	case n < 0x80:
		return append(out, byte(n))
	case n < 0x100:
		return append(out, 0xcc, byte(n))
	case n < 0x10000:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		return append(out, 0xcd, b[0], b[1])
	case n < 0x100000000:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		return append(append(out, 0xce), b[:]...)
	default:
		var b [8]byte
		binary.BigEndian.PutUint64(b[:], n)
		return append(append(out, 0xcf), b[:]...)
	}
}

func mpEncodeFloat(out []byte, f float64) []byte {
	var b [8]byte
	binary.BigEndian.PutUint64(b[:], math.Float64bits(f))
	return append(append(out, 0xcb), b[:]...)
}

func mpEncodeString(out []byte, s string) []byte {
	n := len(s)
	switch {
	case n < 0x20:
		out = append(out, 0xa0|byte(n))
	case n < 0x100:
		out = append(out, 0xd9, byte(n))
	case n < 0x10000:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		out = append(out, 0xda, b[0], b[1])
	default:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		out = append(append(out, 0xdb), b[:]...)
	}
	return append(out, s...)
}

func mpEncodeBinary(out []byte, data []byte) []byte {
	n := len(data)
	switch {
	case n < 0x100:
		out = append(out, 0xc4, byte(n))
	case n < 0x10000:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		out = append(out, 0xc5, b[0], b[1])
	default:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		out = append(append(out, 0xc6), b[:]...)
	}
	return append(out, data...)
}

func mpEncodeArray(out []byte, arr []any) []byte {
	n := len(arr)
	switch {
	case n < 0x10:
		out = append(out, 0x90|byte(n))
	case n < 0x10000:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		out = append(out, 0xdc, b[0], b[1])
	default:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		out = append(append(out, 0xdd), b[:]...)
	}
	for _, item := range arr {
		out = mpEncodeValue(out, item)
	}
	return out
}

func mpEncodeMap(out []byte, m map[string]any) []byte {
	n := len(m)
	switch {
	case n < 0x10:
		out = append(out, 0x80|byte(n))
	case n < 0x10000:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		out = append(out, 0xde, b[0], b[1])
	default:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		out = append(append(out, 0xdf), b[:]...)
	}
	for k, v := range m {
		out = mpEncodeString(out, k)
		out = mpEncodeValue(out, v)
	}
	return out
}

// ---- decode ---------------------------------------------------------------

type mpReader struct {
	buf []byte
	pos int
}

func mpDecode(data []byte) (v any, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("msgpack: malformed input: %v", r)
		}
	}()
	r := &mpReader{buf: data}
	return r.value(), nil
}

func (r *mpReader) byte() byte {
	b := r.buf[r.pos]
	r.pos++
	return b
}

func (r *mpReader) take(n int) []byte {
	s := r.buf[r.pos : r.pos+n]
	r.pos += n
	return s
}

func (r *mpReader) value() any {
	b := r.byte()
	switch {
	case b <= 0x7f:
		return int64(b)
	case b >= 0xe0:
		return int64(int8(b))
	case b >= 0x80 && b <= 0x8f:
		return r.mapOf(int(b & 0x0f))
	case b >= 0x90 && b <= 0x9f:
		return r.arrayOf(int(b & 0x0f))
	case b >= 0xa0 && b <= 0xbf:
		return string(r.take(int(b & 0x1f)))
	}
	switch b {
	case 0xc0:
		return nil
	case 0xc2:
		return false
	case 0xc3:
		return true
	case 0xc4:
		return append([]byte(nil), r.take(int(r.byte()))...)
	case 0xc5:
		return append([]byte(nil), r.take(int(binary.BigEndian.Uint16(r.take(2))))...)
	case 0xc6:
		return append([]byte(nil), r.take(int(binary.BigEndian.Uint32(r.take(4))))...)
	case 0xca:
		return float64(math.Float32frombits(binary.BigEndian.Uint32(r.take(4))))
	case 0xcb:
		return math.Float64frombits(binary.BigEndian.Uint64(r.take(8)))
	case 0xcc:
		return int64(r.byte())
	case 0xcd:
		return int64(binary.BigEndian.Uint16(r.take(2)))
	case 0xce:
		return int64(binary.BigEndian.Uint32(r.take(4)))
	case 0xcf:
		u := binary.BigEndian.Uint64(r.take(8))
		if u <= math.MaxInt64 {
			return int64(u)
		}
		return u
	case 0xd0:
		return int64(int8(r.byte()))
	case 0xd1:
		return int64(int16(binary.BigEndian.Uint16(r.take(2))))
	case 0xd2:
		return int64(int32(binary.BigEndian.Uint32(r.take(4))))
	case 0xd3:
		return int64(binary.BigEndian.Uint64(r.take(8)))
	case 0xd9:
		return string(r.take(int(r.byte())))
	case 0xda:
		return string(r.take(int(binary.BigEndian.Uint16(r.take(2)))))
	case 0xdb:
		return string(r.take(int(binary.BigEndian.Uint32(r.take(4)))))
	case 0xdc:
		return r.arrayOf(int(binary.BigEndian.Uint16(r.take(2))))
	case 0xdd:
		return r.arrayOf(int(binary.BigEndian.Uint32(r.take(4))))
	case 0xde:
		return r.mapOf(int(binary.BigEndian.Uint16(r.take(2))))
	case 0xdf:
		return r.mapOf(int(binary.BigEndian.Uint32(r.take(4))))
	default:
		panic(fmt.Sprintf("unsupported type byte 0x%x", b))
	}
}

func (r *mpReader) arrayOf(n int) []any {
	arr := make([]any, n)
	for i := 0; i < n; i++ {
		arr[i] = r.value()
	}
	return arr
}

func (r *mpReader) mapOf(n int) map[string]any {
	m := make(map[string]any, n)
	for i := 0; i < n; i++ {
		key := r.value()
		val := r.value()
		ks, _ := key.(string)
		m[ks] = val
	}
	return m
}
